lib/utils/compose.js

```diff
@@ -0,0 +1,1086 @@
+				.then(tarStream =>
+					makeBuildTasks(
+						composition,
+						tarStream,
+						{ arch, deviceType },
+						logger,
+						projectName,
+					),
+				)
+				.map(function(task) {
+					const d = imageDescriptorsByServiceName[task.serviceName];
+
+					// multibuild parses the composition internally so any tags we've
+					// set before are lost; re-assign them here
+					if (task.tag == null) {
+						task.tag = [projectName, task.serviceName].join('_').toLowerCase();
+					}
+					if (d.image.context != null) {
+						d.image.tag = task.tag;
+					}
+
+					// configure build opts appropriately
+					if (task.dockerOpts == null) {
+						task.dockerOpts = {};
+					}
+					_.merge(task.dockerOpts, buildOpts, { t: task.tag });
+					if (d.image.context?.args != null) {
+						if (task.dockerOpts.buildargs == null) {
+							task.dockerOpts.buildargs = {};
+						}
+						_.merge(task.dockerOpts.buildargs, d.image.context.args);
+					}
+
+					// Get the service-specific log stream
+					// Caveat: `multibuild.BuildTask` defines no `logStream` property
+					// but it's convenient to store it there; it's JS ultimately.
+					task.logStream = renderer.streams[task.serviceName];
+					task.logBuffer = [];
+
+					// Setup emulation if needed
+					if (task.external || !needsQemu) {
+						return [task, null];
+					}
+					const binPath = qemu.qemuPathInContext(
+						path.join(projectPath, task.context),
+					);
+					return transpose
+						.transposeTarStream(
+							task.buildStream,
+							{
+								hostQemuPath: toPosixPath(binPath),
+								containerQemuPath: `/tmp/${qemu.QEMU_BIN_NAME}`,
+								qemuFileMode: 0o555,
+							},
+							dockerfilePath || undefined,
+						)
+						.then(stream => {
+							task.buildStream = stream;
+						})
+						.return([task, binPath]);
+				}),
+		)
+		.map(function([task, qemuPath]) {
+			const captureStream = buildLogCapture(task.external, task.logBuffer);
+
+			if (task.external) {
+				// External image -- there's no build to be performed,
+				// just follow pull progress.
+				captureStream.pipe(task.logStream);
+				task.progressHook = pullProgressAdapter(captureStream);
+			} else {
+				task.streamHook = function(stream) {
+					let rawStream;
+					stream = createLogStream(stream);
+					if (qemuPath != null) {
+						const buildThroughStream = transpose.getBuildThroughStream({
+							hostQemuPath: toPosixPath(qemuPath),
+							containerQemuPath: `/tmp/${qemu.QEMU_BIN_NAME}`,
+						});
+						rawStream = stream.pipe(buildThroughStream);
+					} else {
+						rawStream = stream;
+					}
+					// `stream` sends out raw strings in contrast to `task.progressHook`
+					// where we're given objects. capture these strings as they come
+					// before we parse them.
+					return rawStream
+						.pipe(dropEmptyLinesStream())
+						.pipe(captureStream)
+						.pipe(buildProgressAdapter(inlineLogs))
+						.pipe(task.logStream);
+				};
+			}
+			return task;
```

Comment:
This used to do
```js
return Promise.resolve(task).tap(function(task) {
```
but since it's all synchronous I changed to this simpler return at the end. It also avoids a linting issue with shadowing `task` that the previous version had